velox/expression/tests/ArgGenerator.cpp

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/expression/tests/ArgGenerator.h"
+#include <boost/random/uniform_int_distribution.hpp>
+
+namespace facebook::velox::test {
+namespace {
+int32_t rand32(FuzzerGenerator& rng) {
+  return boost::random::uniform_int_distribution<uint32_t>()(rng);
+}
+} // namespace
+
+std::vector<TypePtr> DecimalArgGeneratorBase::generateArgs(
+    const exec::FunctionSignature& /*signature*/,
+    const TypePtr& returnType,
+    FuzzerGenerator& rng) {
+  auto inputs = findInputs(returnType, rng);
+  for (const auto& input : inputs) {
+    if (input == nullptr) {
+      return {};
+    }
+  }
+  return inputs;
+}
+
+std::vector<TypePtr> DecimalArgGeneratorBase::findInputs(
+    const TypePtr& returnType,
+    FuzzerGenerator& rng) const {
+  const auto [p, s] = getDecimalPrecisionScale(*returnType);
+  const auto it = inputs_.find({p, s});
+  if (it == inputs_.end()) {
+    LOG(ERROR) << "Cannot find input types for " << returnType->toString();
+    return {};
+  }
+
+  const auto index = rand32(rng) % it->second.size();
+  return it->second[index];
+}
+
+const std::vector<TypePtr>& DecimalArgGeneratorBase::getAllTypes() const {
```

Comment:
This can be a free function in anonymous namespace. Doesn't need to be a method.